pkgtools/pkgcvsupdate/files/pkgcvsupdate.1
.\"	$NetBSD: pkgcvsupdate.1,v 1.3 2002/01/07 12:39:25 seb Exp $
.\"
.\" Copyright (c) 2001 Stoned Elipot. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Stoned Elipot
.\"	for the NetBSD Project
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd November 19, 2001
.Dt PKGCVSUPDATE 1
.Os
.Sh NAME
.Nm pkgcvsupdate
.Nd run CVS to update package sources in pkgsrc
.Sh SYNOPSIS
.Nm
.Ar arguments
.Sh DESCRIPTION
The
.Nm
utility should be run from a package's main directory, a package category
directory or the top-level pkgsrc directory. It first updates the pkgsrc
infrastructure in
.Pa ${PKGSRCDIR}/mk ,
then the package's files and recursively all the packages' files it
depend. When run from the top-level pkgsrc directory it run a separate CVS
update for each category directory, this is believed to be faster than
a single CVS invocation.
.Pp
Any arguments of
.Nm
are passed down to the CVS's update command.
.Sh ENVIRONMENT VARIABLES
.Bl -tag -width CVS_OPTIONS
.It Ev CVS_OPTIONS
Specify options to pass to cvs.
.El
.Sh SEE ALSO
.Xr cvs 1 ,
.Pa ${PKGSRCDIR}/Packages.txt
and
.Rs
.%T "Documentation on the NetBSD Package System"
.Re
.Sh AUTHORS
The pkgcvsupdate utility was written by
.An Stoned Elipot Aq seb@netbsd.org .


doc/NEWS
# $NetBSD: NEWS,v 1.33 2025/12/03 21:57:26 leot Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

m68k softfloat (m68ksf) support [nat 2025-10-22]

        Support for m68k softfloat was added to pkgsrc.

Default versions bumped [wiz 2025-10-06]

	PYTHON_VERSION_DEFAULT is now 313.

Default MySQL implementation changed to MariaDB 11.4 [nia 2024-12-25]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.11 to MariaDB 11.4.

Default versions bumped [wiz 2024-08-12]

	PYTHON_VERSION_DEFAULT is now 312.

Default versions bumped [taca 2024-02-10]

	RUBY_VERSION_DEFAULT is now 32.

Default MySQL implementation changed to MariaDB 10.11 [nia 2024-01-23]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.6 to MariaDB 10.11.

Default versions bumped [wiz 2023-08-14]

	PYTHON_VERSION_DEFAULT is now 311.

Support for USE_CC_FEATURES and USE_CXX_FEAUTRES [nia 2023-07-18]

	USE_CC_FEATURES and USE_CXX_FEATURES can now be used
	to specify a minimum level of C and C++ support required
	to compile a package, including specific libstdc++ features.

	ALLOW_NEWER_COMPILER can be set to no in mk.conf to skip
	building packages that require a newer compiler than the system
	provides.

Default versions bumped [nia 2022-06-30]

	PYTHON_VERSION_DEFAULT is now 310.
	PGSQL_VERSION_DEFAULT is now 14.
	LUA_VERSION_DEFAULT is now 54.

Dropped support for compiling glib2-using packages without introspection
support [wiz 2022-06-07]

PKGSRC_USE_MKTOOLS support added [jperkin 2022-06-07]

        Added C-based tools to speed up various parts of the
        mk infrastructure via PKGSRC_USE_MKTOOLS=yes in mk.conf.

        For packages with large numbers of dependencies, you should see
        anywhere up to around 10x speed-up for the 'wrapper' phase, and
        perhaps more importantly a huge reduction in system time usage.

Linux infrastructure adapted to non-standard distributions [nia 2022-05-01]

        The Linux pkgsrc infrastructure was adapted to support Linux
        distributions where binaries and shared objects are not
        stored in a normal Unix-like file system layout.  It's hoped
        that this will be useful for pkgsrc development on platforms
        like NixOS and Gobo Linux.

UnixWare support [nia 2021-12-26]

        Support for SCO/Xinuos UnixWare was improved to the point where
        it is possible to complete a bootstrap on this platform.

SHA1 distfile hashes deprecated [nia 2021-10-07]

        Distfile hash algorithms have been changed, keeping SHA512,
        adding BLAKE2s, and removing SHA1 and RMD160.  New packages
        and updates will only generate SHA512 and BLAKE2s, and the
        stored hashes have been adjusted.

Default Python version changed to 3.9 [nia 2021-10-01]

Increased enabled-by-default hardening options [nia 2021-09-27]

	PKGSRC_USE_MKPIE was enabled on some architectures.
	Packages that provide static libraries to other packages
	should be rebuilt with MKPIE enabled.
	Some packages may experience runtime errors on platforms
	like i386 - usually this is due to unsafe assembly code,
	which should be disabled when MKPIE is enabled.

	PKGSRC_USE_SSP was bumped from "yes" to "strong".  This
	is not expected to have a noticable effect.

	More information:
	https://www.NetBSD.org/docs/pkgsrc/hardening.html
	(or see "Security hardening" section of doc/pkgsrc.txt)

Default MySQL implementation changed to MariaDB 10.6 [nia 2021-09-27]

Default PostgreSQL changed to 13 [nia 2021-06-28]

Default MySQL implementation changed to MariaDB [nia 2021-06-24]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MySQL 5.7 to MariaDB 10.5.

	This is configurable by setting MYSQL_VERSION_DEFAULT in mk.conf.

	PHP packages are no longer tied to a specific MySQL version
	due to the native driver being enabled by default.

Python 3.8 default [nia 2020-12-04]
	The default Python version was switched to 3.8.

Default database location changed for NetBSD [wiz 2020-12-02]
	The database of installed packages now defaults to ${PREFIX}/pkgdb
	instead of /var/db/pkg on NetBSD. This matches the default on
	other operating systems and cases where a non-default bootstrap
	is used.

	To deal with this change, see this blog post:

	https://pkgsrc.org/news/pkgdb-change/

	Since this change affects the packaging tools only, there is no change
	to binary packages and they do not need to be rebuilt.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"

pkgtools/pkgcvsupdate/files/pkgcvsupdate.sh
#! /bin/sh -
# $NetBSD: pkgcvsupdate.sh,v 1.3 2002/01/07 12:39:25 seb Exp $

AWK=@AWK@
BASENAME=@BASENAME@
DIRNAME=@DIRNAME@
GREP=@GREP@
LS=@LS@
MAKE=@MAKE@
SED=@SED@
TEST=@TEST@

cvs_cmd_args="$@"
cwd=`pwd`
dirs_updated=":"

# low-level cvs update
cvs_update()
{
_cwd=`pwd`
if $TEST "x$1" = x. ; then
	echo "==> $_cwd"
else
	echo "==> $_cwd/$1"
fi
cvs ${CVS_OPTIONS} update $cvs_cmd_args $1 || return 1
}

# update a directory tree
cvs_update_dir()
{
_cwd=`pwd`
case $dirs_updated in
	*:$_cwd:*) return 0;; # already updated
esac
cvs_update . || return 1
dirs_updated="$dirs_updated$_cwd:"
}

# get the file included in ./Makefile
get_make_included_files()
{
$GREP '^\.include' Makefile | \
$AWK '$2 !~ /mk\//{ gsub(/["<>]/, "", $2); print $2 }' || exit 1
}

# get the "make sub-directory"
get_subdirs()
{
# update local files first
_oldcvs_cmd_args="$cvs_cmd_args"
cvs_cmd_args="-l $cvs_cmd_args"
cvs_update . || return 1
cvs_cmd_args="$oldcvs_cmd_args"
subdirs=`$MAKE -V SUBDIR`
if $TEST $? -ne 0 || $TEST x"$subdirs" = x ; then
	# perhaps we don't have the -V flags to make (pre NetBSD 1.3
	# make ?)
	# XXX FIXME: we could miss a new directory here
	subdirs=""
	for _f in * ; do
		if $TEST -f $_f/Makefile ; then
			subdirs="$subdirs $_f"
		fi
	done
fi
}

# update a list of package directories and their dependencies
recursive_cvs_update_pkgdir()
{
_dirs_toupdate="$*"
while true ; do
	# get the next package directory to update
	set blah $_dirs_toupdate
	shift
	_dir=$1
	if $TEST x"$_dir" = x ; then
		break # all done
	fi
	shift
	_dirs_toupdate="$*"
	# cd into it and update...
	if $TEST -d $_dir ; then 
		cd $_dir || exit 1
		cvs_update_dir
	else
		# ...the directory does not exist, that would be a brand new
		# package we try to check it out instead
		cd `$DIRNAME $_dir` || exit 1
		cvs ${CVS_OPTIONS} checkout $cvs_cmd_args `$BASENAME $_dir` || exit 1
		cd $_dir || exit 1
	fi
	_cwd=`pwd`
	# get the dependencies...
	_depends=`$MAKE show-depends-dirs`
	if $TEST $? -ne 0; then
		# ...it failed try to see if it's because included Makefile
		# fragments are missing and queue up the packages providing
		# them
		_included_files=`get_make_included_files`
		for _f in $_included_files ; do
			case x$_f in
				x../*)
					_dirs_toupdate="$_cwd/`$DIRNAME $_f` $_dirs_toupdate";;	
			esac
		done
		# and we would like to visit again the current package
		# directory in the end
		dirs_updated=`echo $dirs_updated | $SED -e "s,:$_cwd:,:,"`
		_dirs_toupdate="$_dirs_toupdate $_cwd"
	else
		# queue up the dependencies of the current package
		for _p in $_depends ; do
			_dirs_toupdate="$_cwd/../../$_p $_dirs_toupdate"
		done
	fi
done
}

if $TEST -f ../../Packages.txt && $TEST -d ../../pkgtools/digest ; then
	# we are in a package directory...

	# update mk
	cd ../../mk && cvs_update_dir && cd $cwd || exit 1

	# update the current package directory and needed dependencies
	recursive_cvs_update_pkgdir $cwd
	
elif $TEST -f Makefile && \
	( $TEST -f ../Packages.txt || $TEST -f Packages.txt) ; then
	# we are in category directory or in the top-level pkgsrc directory

	# update mk
	$TEST -d ../mk && cd ../mk || cd mk || exit 1
	cvs_update_dir && cd $cwd || exit 1

	# get the packages in this category or all the categories
	get_subdirs
	if $TEST $? -ne 0 || $TEST x"$subdirs" = x ; then
		echo "could not get the sub-directories in $cwd" 1>&2
		exit 1
	fi

	if $TEST -f ../Packages.txt ; then
		# update the category packages' directories and their
		# dependencies
		_dirs=""
		for _p in $subdirs ; do
			_dirs="$_dirs $cwd/$_p"
		done
		recursive_cvs_update_pkgdir $_dirs
	else
		for _p in $subdirs ; do
			cvs_update $_p || exit 1
		done
	fi
else
	echo "Not in a package, category or toplevel directory" 1>&2
	exit 1
fi
